### PR TITLE
OH2-183 Create api of Exam Request

### PR DESCRIPTION
**Pull request by ArnaudFonzam**
See OH2-183.

### CODE REVIEW

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 1]
```diff
@@ -64,6 +66,8 @@
 @RestController
 @Api(value = "/laboratories", produces = MediaType.APPLICATION_JSON_VALUE, authorizations = { @Authorization(value = "apiKey") })
 public class LaboratoryController {
+	
+	private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(OpdController.class);
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:70]:
```suggestion
	private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(LaboratoryController.class);
```

Comment (dbmalkovsky) [reply src/main/java/org/isf/lab/rest/LaboratoryController.java:70]: wrong class name; and remove the import for it at line 42

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 2]
```diff
@@ -315,7 +384,11 @@ public ResponseEntity<List<String>> getMaterials() throws OHServiceException {
 	public ResponseEntity<List<LabWithRowsDTO>> getLaboratoryForPrint(@RequestParam(required = false, defaultValue = "") String examName,
 					@RequestParam(value = "dateFrom") String dateFrom, @RequestParam(value = "dateTo") String dateTo,
 					@RequestParam(value = "patientCode", required = false, defaultValue = "0") int patientCode) throws OHServiceException {
-
+		LOGGER.info("Get lawithRow within specified date: {}");
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:387]: `lawithRow` is not a word; should it be `labWithRow`>

File: src/test/java/org/isf/opd/rest/OpdControllerTest.java [hunk 3]
```diff
@@ -46,6 +46,7 @@
 import org.mockito.MockitoAnnotations;
 import org.modelmapper.ModelMapper;
 import org.slf4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/opd/rest/OpdControllerTest.java:49]: Is that import really needed?

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 4]
```diff
@@ -224,6 +224,9 @@ public ResponseEntity<Boolean> updateLaboratory(@PathVariable Integer code, @Req
 		if (labRow != null) {
 			labRows = new ArrayList<>(labRow);
 		}
+		if(!laboratoryDTO.getResult().equals("")) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:227]:
```suggestion
		if (!laboratoryDTO.getResult().equals("")) {
```

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 5]
```diff
@@ -332,23 +335,21 @@ public ResponseEntity<List<LaboratoryDTO>> getLaboratoryExamRequest(@PathVariabl
...
 	}
 	
 	@GetMapping(value = "/laboratories/examrequest", produces = MediaType.APPLICATION_JSON_VALUE)
 	public ResponseEntity<List<LaboratoryDTO>> getLaboratoryExamRequest() throws OHServiceException {
 		LOGGER.info("Get all Exam Requested: {}");
-		List<Laboratory> labList = laboratoryManager.getLaboratory();
+		List<Laboratory> labList = laboratoryManager.getLaboratory().stream().filter(e -> e.getStatus().equals(LaboratorySTATUS.DRAFT.toString())).collect(Collectors.toList());
 		if (labList == null || labList.isEmpty()) {
 			return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null);
 		} else {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:352]: There really isn't a need to have an `else` when the previous statement is `return`.

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 6]
```diff
@@ -332,23 +335,21 @@ public ResponseEntity<List<LaboratoryDTO>> getLaboratoryExamRequest(@PathVariabl
 			throw new OHAPIException(new OHExceptionMessage(null, "Patient not found!", OHSeverityLevel.ERROR));
 		}
 
-		List<Laboratory> labList = laboratoryManager.getLaboratory(patient);
+		List<Laboratory> labList = laboratoryManager.getLaboratory(patient).stream().filter(e -> e.getStatus().equals(LaboratorySTATUS.DRAFT.toString())).collect(Collectors.toList());
 		if (labList == null || labList.isEmpty()) {
 			return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null);
 		} else {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:341]: There really isn't a need to have an `else` when the previous statement is `return`.

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 7]
```diff
@@ -294,7 +297,7 @@ public ResponseEntity<List<LabWithRowsDTO>> getLaboratory(@PathVariable Integer
 			throw new OHAPIException(new OHExceptionMessage(null, "Patient not found!", OHSeverityLevel.ERROR));
 		}
 
-		List<Laboratory> labList = laboratoryManager.getLaboratory(patient);
+		List<Laboratory> labList = laboratoryManager.getLaboratory(patient).stream().filter(e -> !e.getStatus().equals(LaboratorySTATUS.DRAFT.toString())).collect(Collectors.toList());
 		if (labList == null || labList.isEmpty()) {
 			return ResponseEntity.status(HttpStatus.NO_CONTENT).body(null);
 		} else {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:303]: There really isn't a need to have an `else` when the previous statement is `return`.

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 8]
```diff
@@ -122,19 +122,19 @@ public ResponseEntity<Boolean> newLaboratory(@RequestBody LabWithRowsDTO labWith
...
-					}
+		List<Laboratory> labList = laboratoryManager.getLaboratory(patient).stream().filter(e -> e.getStatus().equals(LaboratorySTATUS.DRAFT.toString())).collect(Collectors.toList());
+		if (!(labList == null || labList.isEmpty())) {
+			for (Laboratory lab:labList) {
+				if (lab.getExam() == exam) {
+					throw new OHAPIException(new OHExceptionMessage(null, "Exam Request already exist!", OHSeverityLevel.ERROR));
 				}
 			}
 		}
+		if(!laboratoryDTO.getResult().equals("")) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:133]:
```suggestion
		if (!laboratoryDTO.getResult().equals("")) {
```

File: src/main/java/org/isf/lab/dto/LaboratoryDTO.java [hunk 9]
```diff
@@ -63,15 +64,26 @@ public class LaboratoryDTO {
...
     @ApiModelProperty(notes = "Laboratory Patient Age", position = 11)
     private Integer age;
 
     @ApiModelProperty(notes = "Laboratory Patient Sex", example = "M", position = 12)
     private String sex;
+    
+    @ApiModelProperty(notes = "Laboratory status", example = "Draft", position = 13)
+    private LaboratorySTATUS status;
+
+    public LaboratorySTATUS getStatus() {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/dto/LaboratoryDTO.java:78]: The formatting in this section looks wrong.

Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/dto/LaboratoryDTO.java:78]: Why is the variable `LaboratorySTATUS` not `LaboratoryStatus`?   Using uppercase like this seems wrong.

File: src/main/java/org/isf/lab/dto/LaboratorySTATUS.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package org.isf.lab.dto;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/dto/LaboratorySTATUS.java:1]: Missing license header and change the class name as note above.

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 11]
```diff
@@ -95,7 +99,7 @@ public LaboratoryController(LabManager laboratoryManager, PatientBrowserManager
 
 	@PostMapping(value = "/laboratories", produces = MediaType.APPLICATION_JSON_VALUE)
 	public ResponseEntity<Boolean> newLaboratory(@RequestBody LabWithRowsDTO labWithRowsDTO) throws OHServiceException {
-
+		LOGGER.info("store exam with result: {}");
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:102]: where is the variable associated with the LOGGER place holder?

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 12]
```diff
@@ -135,7 +149,7 @@ public ResponseEntity<Boolean> newLaboratory(@RequestBody LabWithRowsDTO labWith
 
 	@PostMapping(value = "/laboratories/insertList", produces = MediaType.APPLICATION_JSON_VALUE)
 	public ResponseEntity<Boolean> newLaboratory2(@RequestBody List<LabWithRowsDTO> labsWithRows) throws OHServiceException {
-
+		LOGGER.info("store List of Exam with result: {}");
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:152]: Again, where is the variable associated with the LOGGER placeholder?

File: src/main/java/org/isf/login/rest/LoginController.java [hunk 13]
```diff
@@ -64,7 +64,7 @@ public class LoginController {
 	private TokenProvider tokenProvider;
 	
 	@Autowired
-        private CustomAuthenticationManager authenticationManager;
+    private CustomAuthenticationManager authenticationManager;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/login/rest/LoginController.java:67]: Formatting looks incorrect

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 14]
```diff
@@ -112,19 +116,29 @@ public ResponseEntity<Boolean> newLaboratory(@RequestBody LabWithRowsDTO labWith
...
 			labRows = new ArrayList<String>(labRow);
 		}
-
-		if (labToInsert.getDate() == null) {
-			labToInsert.setDate(LocalDateTime.now());
+		List<Laboratory> labList = laboratoryManager.getLaboratory(patient).stream().filter(e -> e.getStatus().equals(LaboratorySTATUS.DRAFT.toString())).collect(Collectors.toList());
+		if (!(labList == null || labList.isEmpty())) {
+			for (Laboratory lab:labList) {
+				if (lab.getExam() == exam) {
+					throw new OHAPIException(new OHExceptionMessage(null, "Exam Request already exist!", OHSeverityLevel.ERROR));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:129]:
```suggestion
					throw new OHAPIException(new OHExceptionMessage(null, "Exam Request already exists.", OHSeverityLevel.ERROR));
```

File: src/main/java/org/isf/login/rest/LoginController.java (hunk 13 above)
Comment (ArnaudFonzam) [reply src/main/java/org/isf/login/rest/LoginController.java:67]: i don't have this formatting looks on my side

File: src/main/java/org/isf/lab/dto/LaboratoryDTO.java [hunk 15]
```diff
@@ -24,6 +24,7 @@
 import java.time.LocalDateTime;
 
 import org.isf.exam.dto.ExamDTO;
+import org.isf.patient.dto.PatientSTATUS;
```
Comment (mwithi) [inline src/main/java/org/isf/lab/dto/LaboratoryDTO.java:27]:
```suggestion
import org.isf.lab.model.LaboratoryStatus;
import org.isf.patient.dto.PatientSTATUS;
```

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 16]
```diff
@@ -33,6 +33,7 @@
 import org.isf.lab.dto.LabWithRowsDTO;
 import org.isf.lab.dto.LaboratoryDTO;
 import org.isf.lab.dto.LaboratoryRowDTO;
+import org.isf.lab.dto.LaboratoryStatus;
```
Comment (mwithi) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:36]:
```suggestion
```

Comment (mwithi) [reply src/main/java/org/isf/lab/rest/LaboratoryController.java:36]: and add `import org.isf.lab.model.LaboratoryStatus;` (organize imports)

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 17]
```diff
@@ -204,12 +218,21 @@ public ResponseEntity<Boolean> updateLaboratory(@PathVariable Integer code, @Req
...
 		if (labRow != null) {
 			labRows = new ArrayList<>(labRow);
 		}
-		boolean updated = laboratoryManager.updateLaboratory(labToInsert, labRows);
+		if (!laboratoryDTO.getResult().equals("")) {
+			labToInsert.setStatus(LaboratoryStatus.done.toString());
+		}
+		boolean updated;
+		if (!labToInsert.getStatus().equals(LaboratoryStatus.done.toString())) {
+			updated = laboratoryManager.updateExamRequest(labToInsert);
```
Comment (mwithi) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:232]: please align to the method in core

File: src/main/java/org/isf/lab/dto/LaboratoryStatus.java
[the remark this answers is not in the document]
Comment (ArnaudFonzam) [reply src/main/java/org/isf/lab/dto/LaboratoryStatus.java:1]: no

File: src/main/java/org/isf/lab/dto/LaboratoryDTO.java [hunk 18]
```diff
@@ -63,15 +65,18 @@ public class LaboratoryDTO {
...
-    private String inOutPatient;
+    private PatientSTATUS inOutPatient;
 
     @ApiModelProperty(notes = "Laboratory Patient Age", position = 11)
     private Integer age;
 
     @ApiModelProperty(notes = "Laboratory Patient Sex", example = "M", position = 12)
     private String sex;
+    
+    @ApiModelProperty(notes = "Laboratory status", example = "draft", position = 13)
```
Comment (mwithi) [inline src/main/java/org/isf/lab/dto/LaboratoryDTO.java:76]:
```suggestion
    @ApiModelProperty(notes = "Laboratory status", example = "DRAFT", position = 13)
```

Comment (mwithi) [reply src/main/java/org/isf/lab/dto/LaboratoryDTO.java:76]: also change in ExamDTO:
```
	@ApiModelProperty(notes = "Exam Code", example = "01.01", position = 1)
	private String code;

	@ApiModelProperty(notes = "Exam Description", example = "1.1 HB", position = 2)
```

File: src/main/java/org/isf/lab/dto/LaboratoryDTO.java [hunk 19]
```diff
@@ -55,8 +55,8 @@ public class LaboratoryDTO {
     @ApiModelProperty(notes = "lock", example = "0")
     private int lock;
 
-    @ApiModelProperty(notes = "Laboratory Note", position = 7)
-    private String note;
+    @ApiModelProperty(notes = "Laboratory Note", value = "Note by laboratorist", position = 7)
+	private String note;
```
Comment (mwithi) [inline src/main/java/org/isf/lab/dto/LaboratoryDTO.java:59]:
```suggestion
    private String note;
```

File: src/main/java/org/isf/exam/dto/ExamDTO.java [hunk 20]
```diff
@@ -27,11 +27,11 @@
...
 
-    @ApiModelProperty(notes = "Exam Code", example = "99.99", position = 1)
-    private String code;
+	@ApiModelProperty(notes = "Exam Code", example = "01.01", position = 1)
+	private String code;
 
-    @ApiModelProperty(notes = "Exam Description", example = "99.99 HB", position = 2)
-    private String description;
+	@ApiModelProperty(notes = "Exam Description", example = "1.1 HB", position = 2)
+	private String description;
```
Comment (mwithi) [inline src/main/java/org/isf/exam/dto/ExamDTO.java:34]:
```suggestion
    @ApiModelProperty(notes = "Exam Description", example = "1.1 HB", position = 2)
    private String description;
```

File: src/main/java/org/isf/exam/dto/ExamDTO.java
Comment (mwithi) [inline src/main/java/org/isf/exam/dto/ExamDTO.java:1]: please reformat the whole file

File: src/main/java/org/isf/lab/dto/LaboratoryDTO.java
Comment (mwithi) [inline src/main/java/org/isf/lab/dto/LaboratoryDTO.java:1]: please reformat the whole file

File: src/main/java/org/isf/login/rest/LoginController.java
Comment (mwithi) [inline src/main/java/org/isf/login/rest/LoginController.java:1]: please reformat the whole file

**Review by mwithi — APPROVED** [review]
LGTM. Thanks a lot!

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 21]
```diff
@@ -134,8 +177,9 @@ public ResponseEntity<Boolean> newLaboratory(@RequestBody LabWithRowsDTO labWith
 	}
 
 	@PostMapping(value = "/laboratories/insertList", produces = MediaType.APPLICATION_JSON_VALUE)
-	public ResponseEntity<Boolean> newLaboratory2(@RequestBody List<LabWithRowsDTO> labsWithRows) throws OHServiceException {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:137]: Why is this API camel cased (`insertList`) where the one in line 138 above isn't (`examrequest`)?    Let's be consistent.

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 22]
```diff
@@ -134,8 +177,9 @@ public ResponseEntity<Boolean> newLaboratory(@RequestBody LabWithRowsDTO labWith
...
-
+	public ResponseEntity<Boolean> newLaboratory2(@RequestBody List<LabWithRowsDTO> labsWithRows)
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:180]: Add javadoc?

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 23]
```diff
@@ -216,8 +270,23 @@ public ResponseEntity<Boolean> updateLaboratory(@PathVariable Integer code, @Req
 		return ResponseEntity.ok(true);
 	}
 
+	@PutMapping(value = "/laboratories/examrequest/{code}", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:273]: another example of API not camel cased (`examrequest`)

File: src/main/java/org/isf/lab/rest/LaboratoryController.java [hunk 24]
```diff
@@ -233,75 +302,121 @@ public ResponseEntity<Boolean> deleteExam(@PathVariable Integer code) throws OHS
...
+
+		return ResponseEntity.ok(labList.stream().map(lab -> {
+			LaboratoryDTO laboratoryDTO = laboratoryMapper.map2DTO(lab);
+			laboratoryDTO.setInOutPatient(PatientSTATUS.valueOf(lab.getInOutPatient()));
+			laboratoryDTO.setStatus(LaboratoryStatus.valueOf(lab.getStatus().toUpperCase()));
+			return laboratoryDTO;
+		}).collect(Collectors.toList()));
+	}
+
+	@GetMapping(value = "/laboratories/examrequest", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/rest/LaboratoryController.java:400]: here again `examrequest`

**Review by dbmalkovsky — COMMENTED** [review]
At a minimum the APIs should be consistently named (camel case) and there is lots of Javadoc missing to explain the various APIs.

File: src/main/java/org/isf/lab/rest/LaboratoryController.java (hunk 21 above)
Comment (mwithi) [reply src/main/java/org/isf/lab/rest/LaboratoryController.java:137]: I agree... @ArnaudFonzam can you do it ASAP? I've already merged CORE and GUI sides... Thanks!

Comment (dbmalkovsky) [reply src/main/java/org/isf/lab/rest/LaboratoryController.java:137]: Please check the whole file for this issue; thanks

**Pull request merged**